components/cam/src/physics/rrtmgp/cam_optics.F90

```diff
@@ -356,347 +338,138 @@ end subroutine combine_properties
+      real(r8), intent(in), dimension(:,:,:) :: tau_bnd, ssa_bnd, asm_bnd
+      real(r8), intent(out), dimension(:,:,:) :: tau_gpt, ssa_gpt, asm_gpt
+      logical, dimension(ngpt,ncol,nlev) :: iscloudy
+      real(r8), dimension(ncol,nlev) :: combined_cld
+      integer, parameter :: changeseed = 1  ! how many times to permute random seed
+      integer :: icol, ilev, igpt
+
+      ! Combined snow and cloud fraction
+      combined_cld(1:ncol,1:nlev) = max(cld(1:ncol,1:nlev), &
+                                        cldfsnow(1:ncol,1:nlev))
+      ! Get stochastic subcolumn cloud mask
+      call mcica_subcol_mask(ngpt, ncol, nlev, changeseed, &
+                             pmid(1:ncol,1:nlev), &
+                             combined_cld(1:ncol,1:nlev), &
+                             iscloudy(1:ngpt,1:ncol,1:nlev))
+      ! Generate subcolumns for homogeneous clouds
+      do igpt = 1,ngpt
+         do ilev = 1,nlev
+            do icol = 1,ncol
+               if (iscloudy(igpt,icol,ilev)) then
+                  tau_gpt(icol,ilev,igpt) = tau_bnd(icol,ilev,gpt2bnd(igpt))
+                  ssa_gpt(icol,ilev,igpt) = ssa_bnd(icol,ilev,gpt2bnd(igpt))
+                  asm_gpt(icol,ilev,igpt) = asm_bnd(icol,ilev,gpt2bnd(igpt))
+               else
+                  tau_gpt(icol,ilev,igpt) = 0
+                  ssa_gpt(icol,ilev,igpt) = 1
+                  asm_gpt(icol,ilev,igpt) = 0
+               end if
+            end do
+         end do
+      end do
+   end subroutine sample_cloud_optics_sw
 
-      type(physics_state), intent(in) :: state
-      type(physics_buffer_desc), pointer :: pbuf(:)
-      integer, intent(in) :: day_indices(:)
-      type(ty_gas_optics_rrtmgp), intent(in) :: kdist
-      type(ty_optical_props_2str), intent(inout) :: optics_out
-
-      ! Type to hold optics on CAM grid
-      type(cam_optics_type) :: optics_cam
-
-      ! Pointer to cloud fraction on physics buffer
-      real(r8), pointer :: cloud_fraction(:,:), snow_fraction(:,:)
-
-      ! Combined cloud and snow fraction
-      real(r8) :: combined_cloud_fraction(pcols,pver)
-
-      ! For MCICA sampling routine
-      integer, parameter :: changeseed = 1
-
-      ! Dimension sizes
-      integer :: ngpt, ncol, nday, nlay
-
-      ! McICA subcolumn cloud flag
-      logical, allocatable :: iscloudy(:,:,:)
-
-      ! Loop variables
-      integer :: icol, ilev, igpt, iband, iday, ilev_cam, ilev_rad
-
-      ! Set a name for this subroutine to write to error messages
-      character(len=32) :: subname = 'set_cloud_optics_sw'
-
-      ncol = state%ncol
-      ngpt = kdist%get_ngpt()
-
-      ! Allocate array to hold subcolumn cloud flag
-      allocate(iscloudy(ngpt,ncol,pver))
+   !----------------------------------------------------------------------------
 
-      ! Initialize output cloud optics object
+   subroutine compress_optics_sw(day_indices, tau, ssa, asm, tau_day, ssa_day, asm_day)
```

Comment:
Might be good to add a description here of what "compress" means.

Reply:
Done.